### PR TITLE
Check that MPI is not finalized before freeing the communicator

### LINKED ISSUE #346: Document that the lifetime of the DistributedSearchTree should be within MPI initialization region
In #135 we introduced duplicating of user-provided communicators to cleanly separate the library's communication from that of the user. We store the duplicated communicator inside `ArborX::DistributedSearchTree`. However, what we did not realize was that it can lead to the problem in the following situation:
```c++
MPI_Initialize(argc, argv);
ArborX::DistributedSearchTree<DeviceType> tree(comm, primitives);
tree.query(...);
MPI_Finalize();
```
The issue is that the destructor for the distributed tree, which contains `MPI_comm_free(&_comm);` will be called **after** `MPI_Finalize()`.

There is currently nothing in the interface to indicate that it cannot/should not be used this way. After searching online, I cannot say definitively what `MPI_Finalize()` does with non-freed communicators. But as it cleans up all MPI context, I think the proper fix would be to check if MPI is still initialized before freeing the communicator.

Reported-by: @mattbement

### PR DESCRIPTION
**Pull request by aprokop**
Fix #346.

This PR still needs a test. A bit tricky as we need an alternative to `utf_main.cpp`.

### CODE REVIEW

File: src/ArborX_DistributedSearchTree.hpp [hunk 1]
```diff
@@ -42,7 +42,16 @@ class DistributedSearchTree
...
-  ~DistributedSearchTree() { MPI_Comm_free(&_comm); }
+  ~DistributedSearchTree()
+  {
+    int mpi_is_finalized;
+    MPI_Finalized(&mpi_is_finalized);
+    if (!mpi_is_finalized)
+    {
+      MPI_Comm_free(&_comm);
+    }
+    _comm = MPI_COMM_NULL;
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/ArborX_DistributedSearchTree.hpp:53]: Not really, for some reason I wanted to follow "set pointer to null" semantics, but it does not make sense in destructor.

**Comment by aprokop** [comment]
>  am tempted to just call it a user error and be done with it

Why? I don't think this is a user error, as user has no visibility into ArborX internals, and there is nothing in the interface to suggest it.

**Comment by dalg24** [comment]
I also think that the right thing to do is just document that the lifetime of the distributed tree cannot exceed MPI environment finalization.

**Comment by aprokop** [comment]
@dalg24 Do you envision any downsides from the proposed solution? If we can fix/work around it, why not do that?

**Comment by aprokop** [comment]
An alternative solution would be to store user's communicator inside the `DistributedSearchTree`, and the `MPI_Comm_dup()` it in place where it's needed (e.g., 2 `queryDispatch()` function and the constructor). This way, we won't need to call anything in the destructor. Would this be an acceptable solution?

**Comment by dalg24** [comment]
The current behavior is **not a defect**, that is to say there is nothing to fix.

We are not going to start checking whether the user called `MPI_Init` in our `Tree::Tree()` constructor nor ensure in `Tree::query` that `MPI_Finalize` has not been called.

Again this is consistent with `Kokkos::initialize` and `Kokkos::finalize`.

**Comment by aprokop** [comment]
OK. I'm uncomfortable doing this because I don't consider this a proper behavior, but I acquiesce.

**Pull request closed without merging**